File: apps/scala/src/main/scala/algorithm/StacksAndQueues.scala
package algorithm

import scala.collection.mutable.ArrayBuffer

/** 第4章 スタックとキュー */

class StackEmptyException extends RuntimeException("スタックは空です")
class StackFullException  extends RuntimeException("スタックは満杯です")
class QueueEmptyException extends RuntimeException("キューは空です")
class QueueFullException  extends RuntimeException("キューは満杯です")

/** 固定長スタック */
class FixedStack[T](capacity: Int):
  private val stk = new Array[Any](capacity)
  private var ptr = 0

  def isEmpty: Boolean = ptr <= 0
  def isFull:  Boolean = ptr >= capacity
  def size:    Int     = ptr
  def getCapacity: Int = capacity

  def push(value: T): Unit =
    if isFull then throw StackFullException()
    stk(ptr) = value
    ptr += 1

  def pop(): T =
    if isEmpty then throw StackEmptyException()
    ptr -= 1
    stk(ptr).asInstanceOf[T]

  def peek(): T =
    if isEmpty then throw StackEmptyException()
    stk(ptr - 1).asInstanceOf[T]

  def find(value: T): Int =
    (ptr - 1 to 0 by -1).find(i => stk(i) == value).getOrElse(-1)

  def contains(value: T): Boolean = find(value) != -1

  def count(value: T): Int =
    (0 until ptr).count(i => stk(i) == value)

  def clear(): Unit = ptr = 0

  def dump(): Array[Any] = stk.take(ptr)

/** 固定長キュー（リングバッファ） */
class FixedQueue[T](capacity: Int):
  private val que  = new Array[Any](capacity)
  private var front = 0
  private var rear  = 0
  private var num   = 0

  def isEmpty: Boolean = num <= 0
  def isFull:  Boolean = num >= capacity
  def size:    Int     = num
  def getCapacity: Int = capacity

  def enqueue(value: T): Unit =
    if isFull then throw QueueFullException()
    que(rear) = value
    rear += 1
    num  += 1
    if rear == capacity then rear = 0

  def dequeue(): T =
    if isEmpty then throw QueueEmptyException()
    val value = que(front).asInstanceOf[T]
    front += 1
    num   -= 1
    if front == capacity then front = 0
    value

  def peek(): T =
    if isEmpty then throw QueueEmptyException()
    que(front).asInstanceOf[T]

  def find(value: T): Int =
    (0 until num).find(i => que((i + front) % capacity) == value).getOrElse(-1)

  def contains(value: T): Boolean = find(value) != -1

  def count(value: T): Int =
    (0 until num).count(i => que((i + front) % capacity) == value)

  def clear(): Unit =
    front = 0; rear = 0; num = 0

File: apps/fsharp/Chapter04/StacksAndQueues.fs
module StacksAndQueues

exception StackEmptyException of string
exception StackFullException of string
exception QueueEmptyException of string
exception QueueFullException of string

/// 固定長スタック
type FixedStack<'T when 'T : equality>(capacity: int) =
    let stk = Array.zeroCreate<obj> capacity
    let mutable ptr = 0

    member _.IsEmpty() = ptr <= 0
    member _.IsFull() = ptr >= capacity
    member _.Size() = ptr
    member _.GetCapacity() = capacity

    member _.Push(value: 'T) =
        if ptr >= capacity then raise (StackFullException "スタックは満杯です")
        stk.[ptr] <- box value
        ptr <- ptr + 1

    member _.Pop() : 'T =
        if ptr <= 0 then raise (StackEmptyException "スタックは空です")
        ptr <- ptr - 1
        unbox stk.[ptr]

    member _.Peek() : 'T =
        if ptr <= 0 then raise (StackEmptyException "スタックは空です")
        unbox stk.[ptr - 1]

    member _.Find(value: 'T) =
        let mutable result = -1
        let mutable i = ptr - 1
        while i >= 0 && result = -1 do
            if unbox<'T> stk.[i] = value then result <- i
            i <- i - 1
        result

    member this.Contains(value: 'T) = this.Find(value) <> -1

    member _.Count(value: 'T) =
        let mutable c = 0
        for i in 0..ptr-1 do
            if unbox<'T> stk.[i] = value then c <- c + 1
        c

    member _.Clear() = ptr <- 0

/// 固定長キュー（リングバッファ）
type FixedQueue<'T when 'T : equality>(capacity: int) =
    let que = Array.zeroCreate<obj> capacity
    let mutable front = 0
    let mutable rear = 0
    let mutable num = 0

    member _.IsEmpty() = num <= 0
    member _.IsFull() = num >= capacity
    member _.Size() = num
    member _.GetCapacity() = capacity

    member _.Enque(value: 'T) =
        if num >= capacity then raise (QueueFullException "キューは満杯です")
        que.[rear] <- box value
        rear <- (rear + 1) % capacity
        num <- num + 1

    member _.Deque() : 'T =
        if num <= 0 then raise (QueueEmptyException "キューは空です")
        let value = unbox<'T> que.[front]
        front <- (front + 1) % capacity
        num <- num - 1
        value

    member _.Peek() : 'T =
        if num <= 0 then raise (QueueEmptyException "キューは空です")
        unbox<'T> que.[front]

    member _.Find(value: 'T) =
        let mutable result = -1
        for i in 0..num-1 do
            let idx = (i + front) % capacity
            if unbox<'T> que.[idx] = value && result = -1 then result <- i
        result

    member this.Contains(value: 'T) = this.Find(value) <> -1

    member _.Count(value: 'T) =
        let mutable c = 0
        for i in 0..num-1 do
            let idx = (i + front) % capacity
            if unbox<'T> que.[idx] = value then c <- c + 1
        c

    member _.Clear() =
        front <- 0; rear <- 0; num <- 0
